components/ai_chat/resources/untrusted_conversation_frame/components/assistant_response/memory_tool_event.tsx

```diff
@@ -0,0 +1,145 @@
+import { getLocale } from '$web-common/locale'
+import styles from './memory_tool_event.module.scss'
+import {
+  useUntrustedConversationContext
+} from '../../untrusted_conversation_context'
+import * as Mojom from '../../../common/mojom'
+import '../../../common/strings'
+
+interface Props {
+  toolUseEvent: Mojom.ToolUseEvent
+}
+
+const MemoryToolEvent: React.FC<Props> = ({ toolUseEvent }) => {
+  const context = useUntrustedConversationContext()
+  const [memoryExists, setMemoryExists] = React.useState<boolean | null>(null)
+
+  // Parse the memory content from tool input
+  const memoryContent = React.useMemo(() => {
+    if (!toolUseEvent?.argumentsJson) return ''
+    try {
+      const input = JSON.parse(toolUseEvent.argumentsJson)
+      return input?.memory || ''
+    } catch (e) {
+      return ''
+    }
+  }, [toolUseEvent?.argumentsJson])
+
+  // Check if memory exists using HasMemory API and subscribe to changes
+  React.useEffect(() => {
+    if (!memoryContent) return
+
+    // Initial check
+    context.uiHandler?.hasMemory(memoryContent).then(({ exists }) => {
+      setMemoryExists(exists)
+    })
+
+    // Subscribe to memory changes via UI observer
+    const id = context.uiObserver?.onMemoriesChanged.addListener(
+      (memories: string[]) => {
+        const exists = memories.includes(memoryContent)
+        setMemoryExists(exists)
+      }
+    )
+
+    return () => {
+      context.uiObserver?.removeListener(id)
+    }
+  }, [memoryContent])
+
+  // Non-empty string in textContentBlock indicates an error
+  const hasError = !!toolUseEvent.output?.[0]?.textContentBlock?.text
+
+  const handleUndo = async () => {
+    if (!memoryContent) return
+
+    await context.uiHandler?.deleteMemory(memoryContent)
+  }
+
+  const handleManageAll = () => {
+    context.uiHandler?.openAIChatCustomizationSettings?.()
+  }
+
+  // Don't render if no memory content, tool hasn't completed, or still
+  // checking memory existence initially.
+  if (!memoryContent || !toolUseEvent.output || memoryExists === null) {
+    return null
+  }
+
+  // Handle error state
+  if (hasError) {
+    return (
+      <div
+        className={styles.memoryToolEvent}
+        data-testid='memory-tool-event-error'
+      >
+        <Icon name='database' className={styles.icon} />
+        <span className={styles.textError}>
+          {getLocale(S.CHAT_UI_MEMORY_ERROR_LABEL)}
+          <button
+            className={`${styles.button} ${styles.buttonWithSpacing}`}
+            onClick={handleManageAll}
+            data-testid='memory-manage-button'
+          >
+            {getLocale(S.CHAT_UI_MEMORY_MANAGE_ALL_BUTTON_LABEL)}
+          </button>
+        </span>
+      </div>
+    )
+  }
+
+  // Handle undone state
+  if (!memoryExists) {
+    return (
+      <div
+        className={styles.memoryToolEvent}
+        data-testid='memory-tool-event-undone'
+      >
+        <Icon name='database' className={styles.icon} />
```

Comment:
It's a little strange how we're repeating the structure of the container which isn't changing between the 3 versions. Since it's a small component if there's a good reason then we can ignore it, but otherwise it seems easy to just vary the inner content based on e.g. `!memoryExists` or `hasError`

Reply:
https://github.com/brave/brave-core/pull/30660/commits/77eb850a58f27fa67f4f6b461ce195b14a1f1bed